hammer/client.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2024 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/transparency-dev/trillian-tessera/client"
+	"k8s.io/klog/v2"
+)
+
+// newLogClientsFromFlags returns a fetcher and a writer that will read
+// and write leaves to all logs in the `log_url` flag set.
+func newLogClientsFromFlags() (*roundRobinFetcher, *roundRobinLeafWriter) {
+	if len(logURL) == 0 {
+		klog.Exitf("--log_url must be provided")
+	}
+
+	fetchers := []client.Fetcher{}
+	writers := []httpLeafWriter{}
+	for _, s := range logURL {
+		// url must reference a directory, by definition
+		if !strings.HasSuffix(s, "/") {
+			s += "/"
+		}
+
+		rootURL, err := url.Parse(s)
+		if err != nil {
+			klog.Exitf("Invalid log URL: %v", err)
+		}
+		fetchers = append(fetchers, newFetcher(rootURL))
+		addURL, err := rootURL.Parse("add")
+		if err != nil {
+			klog.Exitf("Failed to create add URL: %v", err)
+		}
+		writers = append(writers, httpLeafWriter{u: addURL})
+	}
+	return &roundRobinFetcher{f: fetchers}, &roundRobinLeafWriter{ws: writers}
+}
+
+// newFetcher creates a Fetcher for the log at the given root location.
+func newFetcher(root *url.URL) client.Fetcher {
+	get := getByScheme[root.Scheme]
+	if get == nil {
+		panic(fmt.Errorf("unsupported URL scheme %s", root.Scheme))
+	}
+
+	return func(ctx context.Context, p string) ([]byte, error) {
+		u, err := root.Parse(p)
+		if err != nil {
+			return nil, err
+		}
+		return get(ctx, u)
+	}
+}
+
+var getByScheme = map[string]func(context.Context, *url.URL) ([]byte, error){
+	"http":  readHTTP,
+	"https": readHTTP,
```

Comment:
Is it worth supporting the `file:///` scheme too for testing POSIX logs? Could "just" run a busybox HTTP server there, but if we want to use this in CI maybe fewer moving parts is better?

Reply:
Agree from a load-test PoV, but this is probably also the only tool we have currently to test for correctness/conformance.

Happy for it to be added later, though.